vumi/middleware/session_length.py

```diff
@@ -0,0 +1,87 @@
+# -*- test-case-name: vumi.middleware.tests.test_session_length -*-
+
+from twisted.internet.defer import inlineCallbacks, returnValue
+from twisted.internet import reactor
+
+from vumi.message import TransportUserMessage
+from vumi.middleware.base import BaseMiddleware
+from vumi.persist.txredis_manager import TxRedisManager
+
+
+class SessionLengthMiddleware(BaseMiddleware):
+    """ Middleware for storing the session length in the message.
+
+    Stores the start timestamp of a session in the message payload in
+    message['helper_metadata'][field_name]['session_start'] for all messages
+    in the session, as well as the end timestamp if the session in
+    message['helper_metadata'][field_name]['session_end'] if the message
+    marks the end of the session.
+
+    Configuration option:
+
+    :param dict redis:
+        Redis configuration parameters.
+    :param int timeout:
+        Redis key timeout (in seconds). Defaults to 120.
+    :param str field_name:
+        The field name to use when storing the timestamps in the message
+        helper_metadata. Defaults to 'session'.
+    """
+    SESSION_NEW, SESSION_CLOSE = (
+        TransportUserMessage.SESSION_NEW, TransportUserMessage.SESSION_CLOSE)
+
+    @inlineCallbacks
+    def setup_middleware(self):
+        r_config = self.config.get('redis_manager', {})
+        self.redis = yield TxRedisManager.from_config(r_config)
+        self.timeout = self.config.get('timeout', 120)
+        self.field_name = self.config.get('field_name', 'session')
+        self.clock = reactor
+
+    @inlineCallbacks
+    def teardown_middleware(self):
+        yield self.redis.close_manager()
+
+    def _set_message_billing_metadata(self, message, field, value):
```

Comment:
Nothing else mentions ``billing`` so perhaps this method should be ``_set_message_session_metadata``?